polyglot/piranha/src/models/rule.rs

```diff
@@ -73,147 +140,148 @@ impl Rule {
-    self.update(&relevant_substitutions)
-  }
 
-  /// Create a new query from `self` by updating the `query` and `replace` based on the substitutions.
-  fn update(&self, substitutions: &HashMap<String, String>) -> Result<Rule, String> {
-    if substitutions.len() != self.holes().len() {
-      #[rustfmt::skip]
-        return Err(format!("Could not instantiate a rule - {:?}. Some Holes {:?} not found in table {:?}", self, self.holes(), substitutions));
-    } else {
-      let mut updated_rule = self.clone();
-      if !updated_rule.holes().is_empty() {
-        updated_rule.update_query(substitutions);
-        if !updated_rule.is_match_only_rule() {
-          updated_rule.update_replace(substitutions);
-        }
-      }
-      Ok(updated_rule)
+    if relevant_substitutions.len() != self.holes().len() {
+      return Err(format!(
+        "Could not instantiate a rule - {:?}. Some Holes {:?} not found in table {:?}",
+        self,
+        self.holes(),
+        substitutions
+      ));
     }
+    Ok(relevant_substitutions)
   }
-
-  /// Groups the rules based on the field `rule.groups`
-  /// Note: a rule can belong to more than one group.
-  pub(crate) fn group_rules(
-    rules: &Vec<Rule>,
-  ) -> (HashMap<String, Rule>, HashMap<String, Vec<String>>) {
-    let mut rules_by_name = HashMap::new();
-    let mut rules_by_group = HashMap::new();
-    for rule in rules {
-      rules_by_name.insert(rule.name.to_string(), rule.clone());
-      for tag in rule.groups() {
-        rules_by_group.collect(tag.to_string(), rule.name.to_string());
-      }
-    }
-    (rules_by_name, rules_by_group)
+  /// Tries to instantiate the rule (`self`) based on the substitutions.
+  /// Note this could fail if the `substitutions` doesn't contain mappings for each hole.
+  pub(crate) fn try_instantiate(
+    &self, substitutions: &HashMap<String, String>,
+  ) -> Result<Rule, String> {
+    self
+      ._get_substitution_for_holes(substitutions)
+      .map(|x| self.apply_substitutions(&x))
   }
 
-  /// Records the string that should be grepped in order to find files that
-  /// potentially could match this global rule.
-  pub(crate) fn add_grep_heuristics_for_global_rules(
-    &mut self, substitutions: &HashMap<String, String>,
-  ) {
-    let mut gh = HashSet::new();
-    for hole in self.holes() {
-      if let Some(x) = substitutions.get(&hole) {
-        gh.insert(x.clone());
-      }
+  fn apply_substitutions(&self, substitutions: &HashMap<String, String>) -> Rule {
+    let update = |value: &String, is_tree_sitter_query: bool| {
+      substitute_tags(value.to_string(), substitutions, is_tree_sitter_query)
+    };
+
+    let gh = HashSet::from_iter(
+      self
+        ._get_substitution_for_holes(substitutions)
+        .unwrap()
+        .values()
+        .map(|x| x.to_string()),
+    );
+
+    match &self {
+      r @ Rule::Rewrite {
+        query,
+        replace,
+        replace_node,
+        ..
+      } => Rule::Rewrite {
```

Comment:
Lol I know this looks funny. Functional Record Update Syntax doesnt work for enums, it only works for struct , thus this redundancy.